source/dub/generators/generator.d

```diff
@@ -143,130 +142,151 @@ class ProjectGenerator
 
-	private BuildSettings collect(GeneratorSettings settings, Package pack, ref TargetInfo[string] targets, in string[string] configs, ref string[] main_files, string bin_pack)
+	private BuildSettings collect(GeneratorSettings settings, Package pack, BuildSettings parentSettings, ref TargetInfo[string] targets, in string[string] configs, ref string[] main_files, string bin_pack, BuildSettings[string] forDependenciesSettingsMap = null)
 	{
 		import std.algorithm : sort;
 		import dub.compilers.utils : isLinkerFile;
+		import std.stdio;
+		import dub.internal.utils : stripDlangSpecialChars;
 
-		if (auto pt = pack.name in targets) return pt.buildSettings;
-
-		// determine the actual target type
-		auto shallowbs = pack.getBuildSettings(settings.platform, configs[pack.name]);
-		TargetType tt = shallowbs.targetType;
-		if (pack is m_project.rootPackage) {
-			if (tt == TargetType.autodetect || tt == TargetType.library) tt = TargetType.staticLibrary;
-		} else {
-			if (tt == TargetType.autodetect || tt == TargetType.library) tt = settings.combined ? TargetType.sourceLibrary : TargetType.staticLibrary;
-			else if (tt == TargetType.dynamicLibrary) {
-				logWarn("Dynamic libraries are not yet supported as dependencies - building as static library.");
-				tt = TargetType.staticLibrary;
-			}
-		}
-		if (tt != TargetType.none && tt != TargetType.sourceLibrary && shallowbs.sourceFiles.empty) {
-			logWarn(`Configuration '%s' of package %s contains no source files. Please add {"targetType": "none"} to its package description to avoid building it.`,
-				configs[pack.name], pack.name);
-			tt = TargetType.none;
-		}
-
-		shallowbs.targetType = tt;
-		bool generates_binary = tt != TargetType.sourceLibrary && tt != TargetType.none;
-		bool is_target = generates_binary || pack is m_project.rootPackage;
+		BuildSettings buildsettings, forDependenciesBuildSettings;
+		bool is_target;
 
-		if (tt == TargetType.none) {
-			// ignore any build settings for targetType none (only dependencies will be processed)
-			shallowbs = BuildSettings.init;
-			shallowbs.targetType = TargetType.none;
+		void setIs_target(TargetType tt)
+		{
+			bool generates_binary = tt != TargetType.sourceLibrary && tt != TargetType.none;
+			is_target = generates_binary || pack is m_project.rootPackage;
 		}
 
-		// start to build up the build settings
-		BuildSettings buildsettings;
-		processVars(buildsettings, m_project, pack, shallowbs, true);
+		static auto mergeFromDependents(BuildSettings parent, ref BuildSettings child) {
+			child.addVersions(parent.versions);
+			child.addDebugVersions(parent.debugVersions);
+			child.addOptions(BuildOptions(cast(BuildOptions)parent.options & inheritedBuildOptions));
+
+			// special support for overriding string imports in parent packages
+			// this is a candidate for deprecation, once an alternative approach
+			// has been found
+			if (child.stringImportPaths.length) {
```

Comment:
Since when has that been? There are large piles of code relying on this. Before removing this, there needs to be some form of replacement established. I wouldn't expect this to be possible earlier than two years from now.

Reply:
Sorry, seems like I remembered this incorrectly or confused the issue. Thought this was a bug, but it's actually handled correctly. Dub determines the highest package for each string import file.
What remains is that this pollutes the string imports. If a parent overrides a single file, dependencies get `-Jparentpath` and can import anything from the parent.